### PR TITLE
Add a new API to get free mem in mempool

### PR DESCRIPTION
**Pull request by no1wudi**

### CODE REVIEW

File: core/iwasm/common/wasm_memory.c [hunk 1]
```diff
@@ -167,3 +167,16 @@ wasm_runtime_free(void *ptr)
 {
     wasm_runtime_free_internal(ptr);
 }
+
+extern void *
+gc_heap_stats(void *heap, uint32 *stats, int size);
```
Comment (wenyongh) [inline core/iwasm/common/wasm_memory.c:172]: Not a good coding custom, had better implement the API in mem alloc layer, how about:
define struct and implement API in core/shared/mem-alloc/mem_alloc.{hc}:
```
typedef struct mem_alloc_info_t {
        uint32_t total_size;
        uint32_t total_free_size;
        uint32_t highmark_size;
} mem_alloc_info_t;

bool
mem_allocator_get_alloc_info(mem_allocator_t allocator, mem_alloc_info_t *mem_alloc_info);
```

File: core/iwasm/common/wasm_memory.c [hunk 2]
```diff
@@ -167,3 +167,16 @@ wasm_runtime_free(void *ptr)
...
+
+int
```
Comment (wenyongh) [inline core/iwasm/common/wasm_memory.c:174]: Had better return bool, like other APIs in wasm_export.h

File: core/iwasm/common/wasm_memory.c [hunk 3]
```diff
@@ -167,3 +167,16 @@ wasm_runtime_free(void *ptr)
...
+wasm_runtime_mallinfo(mem_mallinfo_t *mall)
+{
+    if (memory_mode == MEMORY_MODE_POOL) {
+        gc_heap_stats(pool_allocator, mall->info, 3);
```
Comment (wenyongh) [inline core/iwasm/common/wasm_memory.c:178]: return mem_allocator_get_alloc_info(pool_allocator, mem_alloc_info);

File: core/iwasm/include/wasm_export.h [hunk 4]
```diff
@@ -121,6 +121,18 @@ typedef union MemAllocOption {
...
+typedef union {
+    struct {
+        uint32_t info[3];
+    };
+    struct {
+        uint32_t total_size;
+        uint32_t total_free_size;
+        uint32_t highmark_size;
+    };
+} mem_mallinfo_t;
```
Comment (wenyongh) [inline core/iwasm/include/wasm_export.h:134]: How about:
```
typedef struct mem_alloc_info_t {
        uint32_t total_size;
        uint32_t total_free_size;
        uint32_t highmark_size;
} mem_alloc_info_t;
```

File: core/iwasm/include/wasm_export.h [hunk 5]
```diff
@@ -229,6 +241,12 @@ wasm_runtime_realloc(void *ptr, unsigned int size);
 WASM_RUNTIME_API_EXTERN void
 wasm_runtime_free(void *ptr);
 
+/*
+ * Get memory info, only pool mode supported now.
+ */
+WASM_RUNTIME_API_EXTERN int
+wasm_runtime_mallinfo(mem_mallinfo_t *mall);
```
Comment (wenyongh) [inline core/iwasm/include/wasm_export.h:248]: How about
```C
WASM_RUNTIME_API_EXTERN bool
wasm_runtime_get_mem_alloc_info(mem_alloc_info_t *mem_alloc_info);
```

File: core/iwasm/common/wasm_memory.c (hunk 1 above)
Comment (no1wudi) [reply core/iwasm/common/wasm_memory.c:172]: This typedef is dupped with wasm_export.h

Comment (wenyongh) [reply core/iwasm/common/wasm_memory.c:172]: How about using a macro to control it (like MEM_ALLOC_OPTION_DEFINED): in wasm_export.h and mem_alloc.h, add:
```C
#ifndef MEM_ALLOC_INFO_DEFINED
#define MEM_ALLOC_INFO_DEFINED
typedef struct mem_alloc_info_t {
        uint32_t total_size;
        uint32_t total_free_size;
        uint32_t highmark_size;
} mem_alloc_info_t;
#endif
```

File: core/iwasm/common/wasm_memory.c [hunk 6]
```diff
@@ -6,6 +6,7 @@
 #include "wasm_runtime_common.h"
 #include "bh_platform.h"
 #include "mem_alloc.h"
+#include "ems/ems_gc.h"
```
Comment (wenyongh) [inline core/iwasm/common/wasm_memory.c:9]: No need to include ems_gc.h

File: core/iwasm/common/wasm_memory.c [hunk 7]
```diff
@@ -167,3 +168,13 @@ wasm_runtime_free(void *ptr)
 {
     wasm_runtime_free_internal(ptr);
 }
+
+bool
+wasm_runtime_get_mem_alloc_info(mem_alloc_info_t *mem_alloc_info)
+{
+    if (memory_mode == MEMORY_MODE_POOL) {
+        return mem_allocator_get_alloc_info(pool_allocator, mem_alloc_info);
+        return true;
```
Comment (wenyongh) [inline core/iwasm/common/wasm_memory.c:177]: Remove unused line

File: core/iwasm/include/wasm_export.h [hunk 8]
```diff
@@ -229,6 +236,12 @@ wasm_runtime_realloc(void *ptr, unsigned int size);
 WASM_RUNTIME_API_EXTERN void
 wasm_runtime_free(void *ptr);
 
+/*
+ * Get memory info, only pool mode supported now.
```
Comment (wenyongh) [inline core/iwasm/include/wasm_export.h:240]: Had better change `supported` to `is supported`

**Pull request merged**